src/relay/op/tensor/transform.cc

```diff
@@ -1300,5 +1302,132 @@ RELAY_REGISTER_OP("slice_like")
 .add_type_rel("SliceLike", SliceLikeRel)
 .set_attr<FTVMCompute>("FTVMCompute", SliceLikeCompute<SliceLikeAttrs>);
 
+
+// relay.layout_transform
+std::pair<Layout, Layout> RemoveLeadingReduandantDimensions(
+    const Layout &src_layout, const Layout &dst_layout, size_t keep_size) {
+  // For a broadcast operator, if the input is a 3-dimensional tensor (64, 1, 1),
+  // we can still apply 4-dimensional rule NCHW -> NCHW16c to it.
+  // In this case, we will delete leading redundant dimensions.
+  CHECK_GE(src_layout.ndim(), keep_size)
+    << "Can only apply layout transform rule for smaller tensor dimensions";
+  int n_remove = src_layout.ndim() - keep_size;
+  CHECK_GT(dst_layout.ndim(), n_remove);
+  for (int i = 0; i < n_remove; ++i) {
+    CHECK_EQ(src_layout[i], dst_layout[i])
+      << "Can only delete the same dimension during layout transform";
+    CHECK(Layout::IsSuperdim(src_layout[i]))
+      << "Can only delete a super dimension during layout transform";
+    CHECK_EQ(src_layout.Subsizeof(src_layout[i]), -1)
+      << "Cannot delete a layout dimension with sub_dimension > 0";
+    CHECK_EQ(dst_layout.Subsizeof(dst_layout[i]), -1)
+      << "Cannot delete a layout dimension with sub_dimension > 0";
+  }
+  return std::make_pair(Layout(src_layout.name().substr(n_remove)),
+                        Layout(dst_layout.name().substr(n_remove)));
+}
+
+Array<Tensor> LayoutTransformCompute(const Attrs& attrs,
+                                     const Array<Tensor>& inputs,
+                                     const Type& out_type,
+                                     const Target& target) {
+  const LayoutTransformAttrs *param = attrs.as<LayoutTransformAttrs>();
+  CHECK(param != nullptr);
+
+  Layout src_layout(param->src_layout);
+  Layout dst_layout(param->dst_layout);
+
+  if (src_layout.Equals(dst_layout)) {
+    return Array<Tensor>{ inputs[0] };
+  }
+
+  CHECK(src_layout.defined() && dst_layout.defined())
+    << "cannot convert from/to undefined layout";
+  CHECK(src_layout.Convertible(dst_layout))
+    << "cannot convert from " << param->src_layout << " to " << param->dst_layout;
+
+  std::tie(src_layout, dst_layout) = RemoveLeadingReduandantDimensions(
+      src_layout, dst_layout, inputs[0]->shape.size());
```

Comment:
RemoveLeadingReduandantDimensions has various checker, while it should still compile even though these restriction are not satisfied. e.g., broadcast(N16nCHW, CHW) -> broadcast(NCHW, CHW) will fail, but actually we are able to make it compile.
Also I'm a little worried about that, the approach allows layout and shape mismatch, and the mismatch is handled in layout_transform operator. In some sense, it's like other operators need to know how layout_transform works, then decide whether to infer a mismatch-shape layout. I would say, from the point view of system design, it's not quite good.

Reply:
Good points. I will clean this mismatch inside this pass.